# Case Study 3:Classification of Zoo Animals

**Submitted by: Princy John**

Import all the modules numpy, pandas, matplotlib,seaborn and KNN

In [14]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sb
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.linear_model import LogisticRegression

Read the zoo.csv file into a Pandas DataSet called: df_zoo


In [15]:
df_zoo = pd.read_csv("zoo.csv")
df_zoo

animal_name  hair  feathers  eggs  milk  airborne  aquatic  predator  \
0      aardvark     1         0     0     1         0        0         1   
1      antelope     1         0     0     1         0        0         0   
2          bass     0         0     1     0         0        1         1   
3          bear     1         0     0     1         0        0         1   
4          boar     1         0     0     1         0        0         1   
..          ...   ...       ...   ...   ...       ...      ...       ...   
96      wallaby     1         0     0     1         0        0         0   
97         wasp     1         0     1     0         1        0         0   
98         wolf     1         0     0     1         0        0         1   
99         worm     0         0     1     0         0        0         0   
100        wren     0         1     1     0         1        0         0   

     toothed  backbone  breathes  venomous  fins  legs  tail  domestic  \
0          1         1         1         0     0     4     0         0   
1          1         1         1         0     0     4     1         0   
2          1         1         0         0     1     0     1         0   
3          1         1         1         0     0     4     0         0   
4          1         1         1         0     0     4     1         0   
..       ...       ...       ...       ...   ...   ...   ...       ...   
96         1         1         1         0     0     2     1         0   
97         0         0         1         1     0     6     0         0   
98         1         1         1         0     0     4     1         0   
99         0         0         1         0     0     0     0         0   
100        0         1         1         0     0     2     1         0   

     catsize  class_type  
0          1           1  
1          1           1  
2          0           4  
3          1           1  
4          1           1  
..       ...         ...  
96         1           1  
97         0           6  
98         1           1  
99         0           7  
100        0           2  

[101 rows x 18 columns]

Use the info method to print out some info about df_zoo

In [16]:
df_zoo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101 entries, 0 to 100
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   animal_name  101 non-null    object
 1   hair         101 non-null    int64 
 2   feathers     101 non-null    int64 
 3   eggs         101 non-null    int64 
 4   milk         101 non-null    int64 
 5   airborne     101 non-null    int64 
 6   aquatic      101 non-null    int64 
 7   predator     101 non-null    int64 
 8   toothed      101 non-null    int64 
 9   backbone     101 non-null    int64 
 10  breathes     101 non-null    int64 
 11  venomous     101 non-null    int64 
 12  fins         101 non-null    int64 
 13  legs         101 non-null    int64 
 14  tail         101 non-null    int64 
 15  domestic     101 non-null    int64 
 16  catsize      101 non-null    int64 
 17  class_type   101 non-null    int64 
dtypes: int64(17), object(1)
memory usage: 14.3+ KB


Read the class.csv file into a Pandas DataSet called df_class

In [17]:
df_class = pd.read_csv("class.csv")
df_class

Class_Number  Number_Of_Animal_Species_In_Class    Class_Type  \
0             1                                 41        Mammal   
1             2                                 20          Bird   
2             3                                  5       Reptile   
3             4                                 13          Fish   
4             5                                  4     Amphibian   
5             6                                  8           Bug   
6             7                                 10  Invertebrate   

                                        Animal_Names  
0  aardvark, antelope, bear, boar, buffalo, calf,...  
1  chicken, crow, dove, duck, flamingo, gull, haw...  
2    pitviper, seasnake, slowworm, tortoise, tuatara  
3  bass, carp, catfish, chub, dogfish, haddock, h...  
4                             frog, frog, newt, toad  
5  flea, gnat, honeybee, housefly, ladybird, moth...  
6  clam, crab, crayfish, lobster, octopus, scorpi...

In [18]:
df_class.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column                             Non-Null Count  Dtype 
---  ------                             --------------  ----- 
 0   Class_Number                       7 non-null      int64 
 1   Number_Of_Animal_Species_In_Class  7 non-null      int64 
 2   Class_Type                         7 non-null      object
 3   Animal_Names                       7 non-null      object
dtypes: int64(2), object(2)
memory usage: 352.0+ bytes


Merge the two DataFrames together.

Use the join method of df_zoo, to merge df_class into df_zoo.

Set the index of df_class to be “Class_Number” before or during the join.

When joining use on=”class_type”

In [19]:
df_zoo= df_zoo.join(df_class.set_index("Class_Number"), on="class_type")
df_zoo

animal_name  hair  feathers  eggs  milk  airborne  aquatic  predator  \
0      aardvark     1         0     0     1         0        0         1   
1      antelope     1         0     0     1         0        0         0   
2          bass     0         0     1     0         0        1         1   
3          bear     1         0     0     1         0        0         1   
4          boar     1         0     0     1         0        0         1   
..          ...   ...       ...   ...   ...       ...      ...       ...   
96      wallaby     1         0     0     1         0        0         0   
97         wasp     1         0     1     0         1        0         0   
98         wolf     1         0     0     1         0        0         1   
99         worm     0         0     1     0         0        0         0   
100        wren     0         1     1     0         1        0         0   

     toothed  backbone  ...  venomous  fins  legs  tail  domestic  catsize  \
0          1         1  ...         0     0     4     0         0        1   
1          1         1  ...         0     0     4     1         0        1   
2          1         1  ...         0     1     0     1         0        0   
3          1         1  ...         0     0     4     0         0        1   
4          1         1  ...         0     0     4     1         0        1   
..       ...       ...  ...       ...   ...   ...   ...       ...      ...   
96         1         1  ...         0     0     2     1         0        1   
97         0         0  ...         1     0     6     0         0        0   
98         1         1  ...         0     0     4     1         0        1   
99         0         0  ...         0     0     0     0         0        0   
100        0         1  ...         0     0     2     1         0        0   

     class_type  Number_Of_Animal_Species_In_Class    Class_Type  \
0             1                                 41        Mammal   
1             1                                 41        Mammal   
2             4                                 13          Fish   
3             1                                 41        Mammal   
4             1                                 41        Mammal   
..          ...                                ...           ...   
96            1                                 41        Mammal   
97            6                                  8           Bug   
98            1                                 41        Mammal   
99            7                                 10  Invertebrate   
100           2                                 20          Bird   

                                          Animal_Names  
0    aardvark, antelope, bear, boar, buffalo, calf,...  
1    aardvark, antelope, bear, boar, buffalo, calf,...  
2    bass, carp, catfish, chub, dogfish, haddock, h...  
3    aardvark, antelope, bear, boar, buffalo, calf,...  
4    aardvark, antelope, bear, boar, buffalo, calf,...  
..                                                 ...  
96   aardvark, antelope, bear, boar, buffalo, calf,...  
97   flea, gnat, honeybee, housefly, ladybird, moth...  
98   aardvark, antelope, bear, boar, buffalo, calf,...  
99   clam, crab, crayfish, lobster, octopus, scorpi...  
100  chicken, crow, dove, duck, flamingo, gull, haw...  

[101 rows x 21 columns]

Drop unnecessary columns of the new df_zoo in place:

1. catsize

2. Number_Of_Animal_Species_In_Class

3. Animal_Names

In [20]:
df_zoo.drop(["catsize","Number_Of_Animal_Species_In_Class","Animal_Names"], axis = 1, inplace = True)

Split the data into training and testing data.

Prepare your X and y:

X: Drop the animal_name, class_type, Class_Type columns.
y: Specify the class_type column.

Use sklearn train_test_split to split the data

In [21]:
x = df_zoo.drop(["animal_name", "class_type", "Class_Type"], axis = 1)
y = df_zoo["class_type"]

x_train, x_test, y_train, y_test= train_test_split(x, y, test_size = 0.3)

Create the model and fit it to the training data.

Create an sklearn KNeighborsClassifier model, with k=1.

 Use the fit method to fit it to the training data.

In [22]:
knn = KNeighborsClassifier (n_neighbors = 1)
knn.fit(x_train, y_train)

KNeighborsClassifier(n_neighbors=1)

Predict values based on testing data.

Use the predict method to predict values with the X testing data

In [23]:
pred_value = knn.predict(x_test)

Print out the classification report using y test data and predictions, to see how accurate the model is.

Insert a markdown cell and make some observations about the accuracy of this model.

In [24]:
print(classification_report(y_test,pred_value))

              precision    recall  f1-score   support

           1       1.00      1.00      1.00        10
           2       1.00      1.00      1.00         9
           3       1.00      1.00      1.00         1
           4       1.00      1.00      1.00         2
           5       1.00      1.00      1.00         1
           6       0.80      1.00      0.89         4
           7       1.00      0.75      0.86         4

    accuracy                           0.97        31
   macro avg       0.97      0.96      0.96        31
weighted avg       0.97      0.97      0.97        31



**Observations** 

The accuracy of the data model is 0.94 which is equal to 94%, which is pretty impressive. However, everytime we run the code, the measure of accuracy slightly varies attributing to the randomness of the data being generated. Also, few times the program was run, the accyracy of the model wa 1.0, i.e., 100 per cent.

Repeat steps 10 to 13 using k=3

In [25]:
knn=KNeighborsClassifier(n_neighbors=3)
knn.fit(x_train,y_train)
pred_value=knn.predict(x_test)
print(classification_report(y_test,pred_value))

              precision    recall  f1-score   support

           1       1.00      1.00      1.00        10
           2       1.00      1.00      1.00         9
           3       1.00      1.00      1.00         1
           4       1.00      1.00      1.00         2
           5       1.00      1.00      1.00         1
           6       0.67      1.00      0.80         4
           7       1.00      0.50      0.67         4

    accuracy                           0.94        31
   macro avg       0.95      0.93      0.92        31
weighted avg       0.96      0.94      0.93        31



**Observations:** 
    
    We see that the accuracy of the model is 0.90 and lies between 0.84 and 1.00. Since the value of the k did not increase considerably and because it is only 3, we still find that the accuracy of the data is 100%, sometime.

Try a different machine learning model:

1. create an sklearn LogisticRegression model

2. Fit it to the training data

3. make predictions based on the test data

4. print the classification report

Insert a markdown cell and make some observations about the accuracy of the three different models you tested.

In [26]:
logistic_regression = LogisticRegression().fit(x_train,y_train)

prediction_value =logistic_regression.predict(x_test)

print(classification_report(y_test,prediction_value))


              precision    recall  f1-score   support

           1       1.00      1.00      1.00        10
           2       1.00      1.00      1.00         9
           3       1.00      1.00      1.00         1
           4       1.00      1.00      1.00         2
           5       1.00      1.00      1.00         1
           6       0.67      1.00      0.80         4
           7       1.00      0.50      0.67         4

    accuracy                           0.94        31
   macro avg       0.95      0.93      0.92        31
weighted avg       0.96      0.94      0.93        31



C:\Users\Princy John Simon\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Do some research on the internet to determine in what circumstances you should use a KNN model or LogisticRegression model. Insert a markdown cell and explain what you found. 

Observations: 

Using the Logistic Regression model, the accuracy of the data is never 100%. the accuracy lies in the range of 0.80 to 0.97. That is 80% to 97%. 

Both KNN and Logistic are classic machine leanring algortithms. In the KNN technique, we study and analyse the neighbours and come up with the prediction model. Additonally, this algorithm is based on distance between neighboring points, implying that, it fits and predicts, when we run for the output.

Logistic Regression is based on probability and the bend is fitted to information and the information is grouped according to classes.

Source: https://towardsdatascience.com/comparative-study-on-classic-machine-learning-algorithms-24f9ff6ab222#:~:text=KNN%20is%20a%20non%2Dparametric%20model%2C%20where%20LR%20is%20a,can%20only%20output%20the%20labels